File: Data_Structures/LinkedList/Linked_List_Cycle/FloydCycle.py
#Python program to detect loop in linked list using the Floyd's Algorithm

#node class
class linkedListNode:
    
    #Initializing the node
    def __init__(self,data):
        self.data=data
        self.next= None

#Linked list class
class linkedList:

    #initialize the linked list
    def __init__(self):
        self.head = None

    #function to push data into the linked list
    def push(self, data):
        node = linkedListNode(data)
        node.next = self.head
        self.head = node

    #function to print the linked list
    def printLinkedList(self):
        temp= self.head
        while(temp):
            print(temp.data)
            temp = temp.next

    def loopDetectFloyd(self):
        slow = self.head
        fast = self.head
        while(slow and fast and fast.next):
            slow= slow.next
            fast= fast.next.next
            if slow == fast:
                return True

File: Data_Structures/LinkedList/Linked_List_Cycle/FloydCycle.js
/* Credits to Joshua Blankenship for a Medium article explaining this concept and algorithm in detail
   https://medium.com/@joshuablankenshipnola/checking-for-linked-list-cycles-in-javascript-77ec9adc6822 */

const floydCycle = (linkedList) => {

  // Create a slow pointer and a fast pointer
  let slowPointer = linkedList;
  let fastPointer = linkedList;

  // While both exist and the faster pointer has a next value
  while(fastPointer && fastPointer.next) {
    // Move the slow pointer by a single node
    slowPointer = slowPointer.next;
    // Move the fast pointer by two nodes
    fastPointer = fastPointer.next.next;

    // If both pointers intersect at any point, then the linked list contains a loop
    if (slowPointer === fastPointer) {
      return true;
    }
  }

  // Return false if no cycle is found
  return false;
}